Authenticators/IMAP/imapauth.py

```diff
@@ -0,0 +1,605 @@
+                try:
+                    server.setAuthenticator(app.auth)
+                # Apparently this server was restarted without us noticing
+                except (Murmur.InvalidSecretException, Ice.UnknownUserException) as e:
+                    if hasattr(e, "unknown") and e.unknown != "Murmur::InvalidSecretException":
+                        # Special handling for Murmur 1.2.2 servers with invalid slice files
+                        raise e
+                    
+                    error('Invalid ice secret')
+                    return
+            else:
+                debug('Virtual server %d got started', server.id())
+
+        @fortifyIceFu()
+        @checkSecret
+        def stopped(self, server, current = None):
+            """
+            This function is called when a virtual server is stopped
+            """
+            if self.app.connected:
+                # Only try to output the server id if we think we are still connected to prevent
+                # flooding of our thread pool
+                try:
+                    if not cfg.murmur.servers or server.id() in cfg.murmur.servers:
+                        info('Authenticated virtual server %d got stopped', server.id())
+                    else:
+                        debug('Virtual server %d got stopped', server.id())
+                    return
+                except Ice.ConnectionRefusedException:
+                    self.app.connected = False
+            
+            debug('Server shutdown stopped a virtual server')
+    
+    authenticateFortifyResult = (-2, None, None)
+        
+    class imapauthenticator(Murmur.ServerUpdatingAuthenticator):
+        texture_cache = {}
+        def __init__(self):
+            Murmur.ServerUpdatingAuthenticator.__init__(self)
+
+        @fortifyIceFu(authenticateFortifyResult)
+        @checkSecret
+        def authenticate(self, name, pw, certlist, certhash, strong, current = None):
+            """
+            This function is called to authenticate a user
+            """
+            
+            # Search for the user in the database
+            FALL_THROUGH = -2
+            AUTH_REFUSED = -1
+            
+            if name == 'SuperUser':
+                debug('Forced fall through for SuperUser')
+                return (FALL_THROUGH, None, None)
+            
+
+            try:
+                with IMAPClient(host=cfg.imap.host, ssl_context=ssl_context) as client:
+                    client.login(name, pw)
+            except IMAPClient.Error as e:
+                info('Fall through for unknown user "%s" or connection problem', name)
+                return (AUTH_REFUSED, None, None)
+            
+            # As IMAP hasn't got such thing as a USERID, we use the given name to calculate a hash 
+	    # TODO this might be highly insecure when two mails have coindently the same hash (note that we are cutting the integer off as otherwise we would get an error from mumble)
```

Comment:
As this is a TODO, how would one fix it?

I think this is a technical limitation that is unfixable.

So I would rather put this into the scripts description in the file header as a strong /caveat warning.

Reply:
I think that's better, yes.